src/mapgen_carpathian.cpp

```diff
@@ -0,0 +1,458 @@
+	u32 index3d = 0;
+
+	// Calculate noise for terrain generation
+	noise_base->perlinMap2D(node_min.X, node_min.Z);
+	noise_height1->perlinMap2D(node_min.X, node_min.Z);
+	noise_height2->perlinMap2D(node_min.X, node_min.Z);
+	noise_height3->perlinMap2D(node_min.X, node_min.Z);
+	noise_height4->perlinMap2D(node_min.X, node_min.Z);
+	noise_hills_terrain->perlinMap2D(node_min.X, node_min.Z);
+	noise_ridge_terrain->perlinMap2D(node_min.X, node_min.Z);
+	noise_step_terrain->perlinMap2D(node_min.X, node_min.Z);
+	noise_hills->perlinMap2D(node_min.X, node_min.Z);
+	noise_ridge_mnt->perlinMap2D(node_min.X, node_min.Z);
+	noise_step_mnt->perlinMap2D(node_min.X, node_min.Z);
+	noise_mnt_var->perlinMap3D(node_min.X, node_min.Y - 1, node_min.Z);
+
+	//// Place nodes
+	for (s16 z = node_min.Z; z <= node_max.Z; z++) {
+		for (s16 y = node_min.Y - 1; y <= node_max.Y + 1; y++) {
+			u32 vi = vm->m_area.index(node_min.X, y, z);
+			for (s16 x = node_min.X; x <= node_max.X;
+					x++, vi++, index2d++, index3d++) {
+				if (vm->m_data[vi].getContent() != CONTENT_IGNORE)
+					continue;
+
+				// Base terrain
+				float ground = noise_base->result[index2d];
+
+				// Gradient
+				s32 grad = terrain_zero_level - y;
+				// Shallow seabed
+				if (y < water_level) {
+					grad = grad_wl + (water_level - y) * 3;
+				}
+
+				// Hill/Mountain height (hilliness)
+				float height1 = noise_height1->result[index2d];
+				float height2 = noise_height2->result[index2d];
+				float height3 = noise_height3->result[index2d];
+				float height4 = noise_height4->result[index2d];
+				float mnt_var = noise_mnt_var->result[index3d];
+				// Combine height noises and apply 3D variation
+				float hill1 = getLerp(height1, height2, mnt_var);
+				float hill2 = getLerp(height3, height4, mnt_var);
+				float hill3 = getLerp(height3, height2, mnt_var);
+				float hill4 = getLerp(height1, height4, mnt_var);
+				// 'hilliness' determines whether hills/mountains are
+				// small or large
+				float hilliness = std::fmax(std::fmin(hill1, hill2), std::fmin(hill3, hill4));
+
+				// Rolling hills
+				float hter = noise_hills_terrain->result[index2d];
+				float n_hills = noise_hills->result[index2d];
+				float hill_mnt = hilliness * pow(n_hills, 2.f);
+				float hills = pow(fabs(hter), 3.f) * hill_mnt;
+
+				// Ridged mountains
+				float rter = noise_ridge_terrain->result[index2d];
+				float n_ridge_mnt = noise_ridge_mnt->result[index2d];
+				float ridge_mnt = hilliness * (1.f - fabs(n_ridge_mnt));
+				float ridged_mountains = pow(fabs(rter), 3.f) * ridge_mnt;
+
+				// Step (terraced) mountains
+				float ster = noise_step_terrain->result[index2d];
+				float n_step_mnt = noise_step_mnt->result[index2d];
+				float step_mnt = hilliness * getSteps(n_step_mnt, fabs(mnt_var));
+				float step_mountains = pow(fabs(ster), 3.f) * step_mnt;
+
+				// Final terrain level
+				float mountains = hills + ridged_mountains + step_mountains;
+				float density = ground + mountains + grad;
```

Comment:
Please rename 'density' to 'surface level' or 'surface y' to avoid confusion with 'density mehtod' mapgens which this is not.